ui/components-react/src/components-react/propertygrid/MergingPropertyDataProvider.ts

```diff
@@ -93,12 +110,71 @@ export class MergingPropertyDataProvider implements IPropertyDataProvider {
   }
 
   private mergeCategories(dataArray: PropertyData[]): PropertyCategory[] {
-    const mergedCategories: PropertyCategory[] = [];
+    const mergedCategories = new Array<PropertyCategory>();
     dataArray.forEach((data) => {
-      mergedCategories.push(...data.categories);
+      for (const category of data.categories) {
+        const mergedCategoryIndex = mergedCategories.findIndex(
+          (mergedCat) => mergedCat.name === category.name
+        );
+        if (mergedCategoryIndex === -1) {
+          mergedCategories.push(category);
+          continue;
+        }
+
+        const newChildCategories = this.mergeChildCategories(
+          mergedCategories[mergedCategoryIndex].childCategories,
+          category.childCategories
+        );
+        mergedCategories[mergedCategoryIndex] = {
+          ...mergedCategories[mergedCategoryIndex],
+          ...(newChildCategories
+            ? { childCategories: [...newChildCategories] }
+            : {}),
+        };
+      }
     });
 
-    return mergedCategories;
+    return [...mergedCategories];
+  }
+
+  private mergeChildCategories(
+    lhs: PropertyCategory[] | undefined,
+    rhs: PropertyCategory[] | undefined
+  ): PropertyCategory[] | undefined {
+    if (!lhs || lhs.length === 0) {
+      return rhs;
+    }
+
+    if (!rhs || rhs.length === 0) {
+      return lhs;
+    }
+
+    const mergedCategories = new Array<PropertyCategory>();
+    for (const lhsCategory of lhs) {
+      mergedCategories.push(lhsCategory);
+    }
+
+    for (const rhsCategory of rhs) {
+      const childCategoryIndex = mergedCategories.findIndex(
+        (mergedCategory) => mergedCategory.name === rhsCategory.name
+      );
+      if (childCategoryIndex === -1) {
+        mergedCategories.push(rhsCategory);
+        continue;
+      }
+
+      const newChildCategories = this.mergeChildCategories(
+        mergedCategories[childCategoryIndex].childCategories,
+        rhsCategory.childCategories
+      );
+      mergedCategories[childCategoryIndex] = {
+        ...mergedCategories[childCategoryIndex],
+        ...(newChildCategories
+          ? { childCategories: [...newChildCategories] }
```

Comment:
no need to copy the array

```suggestion
          ? { childCategories: newChildCategories }
```